File: lv1_module2/ros2_ws/src/turtle_cpp/src/distance_publisher.cpp
#include <cmath>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include "turtlesim/msg/pose.hpp"

class DistancePublisher : public rclcpp::Node
{
public:
    DistancePublisher()
    : Node("turtle_distance_publisher"),
      x_(0.0),
      y_(0.0)
    {
        // /turtle1/pose 구독
        pose_subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtle1/pose",
            10,
            std::bind(
                &DistancePublisher::pose_callback,
                this,
                std::placeholders::_1
            )
        );

        // /turtle_distance 발행
        distance_publisher_ =
            this->create_publisher<std_msgs::msg::Float32>(
                "/turtle_distance",
                10
            );

        // 10Hz = 0.1초
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(
                &DistancePublisher::timer_callback,
                this
            )
        );

        RCLCPP_INFO(
            this->get_logger(),
            "거리 발행자 시작 (10 Hz)"
        );
    }

private:
    // /turtle1/pose를 받음
    void pose_callback(
        const turtlesim::msg::Pose::SharedPtr msg)
    {
        x_ = msg->x;
        y_ = msg->y;
    }

    // 10Hz마다 거리 계산 후 발행
    void timer_callback()
    {
        double distance = std::sqrt(
            x_ * x_ + y_ * y_
        );

        std_msgs::msg::Float32 msg;
        msg.data = static_cast<float>(distance);

        distance_publisher_->publish(msg);
    }

    double x_;
    double y_;

    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr
        pose_subscription_;

    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr
        distance_publisher_;

    rclcpp::TimerBase::SharedPtr timer_;
};


int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<DistancePublisher>();

    rclcpp::spin(node);

    rclcpp::shutdown();

    return 0;
}

File: lv1_module2/ros2_ws/src/turtle_py/turtle_py/distance_publisher.py
import math

import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
from std_msgs.msg import Float32
from rcl_interfaces.msg import SetParametersResult


class TurtleDistancePublisher(Node):

    def __init__(self):
        super().__init__('turtle_distance_publisher')

        # 최신 자세
        self.x = 0.0
        self.y = 0.0

        # 파라미터 선언
        self.declare_parameter('publish_rate', 10.0)

        self.publish_rate = self.get_parameter(
            'publish_rate'
        ).value

        # /turtle1/pose 구독
        self.pose_subscription = self.create_subscription(
            Pose,
            '/turtle1/pose',
            self.pose_callback,
            10
        )

        # /turtle_distance 발행
        self.distance_publisher = self.create_publisher(
            Float32,
            '/turtle_distance',
            10
        )

        # Timer
        self.timer = self.create_timer(
            1.0 / self.publish_rate,
            self.timer_callback
        )

        # 파라미터 변경 콜백
        self.add_on_set_parameters_callback(
            self.parameter_callback
        )

    def pose_callback(self, msg):
        self.x = msg.x
        self.y = msg.y

    def timer_callback(self):
        distance = math.sqrt(self.x ** 2 + self.y ** 2)

        msg = Float32()
        msg.data = distance

        self.distance_publisher.publish(msg)

    def parameter_callback(self, params):

        for param in params:

            if param.name == 'publish_rate':

                if param.value <= 0:
                    return SetParametersResult(
                        successful=False,
                        reason='publish_rate must be greater than 0'
                    )

                self.publish_rate = param.value

                # 기존 Timer 제거 후 새로운 주기로 생성
                self.timer.cancel()

                self.timer = self.create_timer(
                    1.0 / self.publish_rate,
                    self.timer_callback
                )

                self.get_logger().info(
                    f'publish_rate changed to '
                    f'{self.publish_rate} Hz'
                )

        return SetParametersResult(successful=True)


def main(args=None):

    # 1. ROS 2 초기화
    rclpy.init(args=args)

    # 2. 노드 생성
    node = TurtleDistancePublisher()

    try:
        # 3. 노드 실행
        rclpy.spin(node)

    except KeyboardInterrupt:
        pass

    finally:
        # 4. 노드 제거
        node.destroy_node()

        # 5. ROS 2 종료
        
        if rclpy.ok():
         rclpy.shutdown()


if __name__ == '__main__':
    main()